Handle absence of value for `--all-or-nothing` properly

agustingomes opened the pull request:
### Expected behavior:
Upon specification of the option `--all-or-nothing` wraps multiple migrations into a transaction

### Observed behavior
Specifying `--all-or-nothing` by itself does not cause the desired effect, as the result of the statement `$input->getOption('all-or-nothing')` is `null`, and therefore, in the absence of config, this value is ultimately evaluated to `false`, therefore not wrapping multiple migrations into a single transaction.


<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | no
| Fixed issues | https://github.com/doctrine/migrations/issues/1279 

#### Summary

<!-- Provide a summary your change. -->

greg0ire commented:
As stated in #1279 today, this looks to me more like an unintended bug introduced in #1161. Can you take a look and let me know what you think?

agustingomes commented:
Hi @greg0ire 

I will try to add a failing test case if possible. I'll keep you posted.

agustingomes commented:
@greg0ire I added an extra permutation for the data provider of the `allOrNothing` test, which confirms that the behavior is not as expected.

What would you advice as next step?

greg0ire commented:
- revert the doc change
- fix the bug: `$allOrNothing   = $input->hasOption('all-or-nothing') ? $input->getOption('all-or-nothing') ?? true : null;` should IMO do the trick.

agustingomes commented:
@greg0ire unfortunately your suggestion alone does not seem to do the trick.

I've checked the [hasOption method in Symfony](https://github.com/symfony/symfony/blob/6.3/src/Symfony/Component/Console/Input/InputDefinition.php#L267) and it seems that method is not a reliable way to know if a user passed the command option or not.

In any case, I'll do a small refactor there to stop using `hasOption`.

greg0ire commented:
Indeed! I think by using a default value different than `null` (like `false`?) for `all-or-nothing`, you might be able to tell whether the option was used or not with `getOption`

agustingomes commented:
I'm not sure @greg0ire 

I'll give it some thought over the week, since it seems like relying on the default value of the option may not work, as it can override the Doctrine project configuration.

greg0ire commented:
I guess if you set the default to `false`, you would then have to translate `null` to `true` (option provided), and `false` to `null` (default value, option not provided), so that the line `$allOrNothing   = (bool) ($allOrNothing ?? $this->configuration->isAllOrNothing());` continues to work as intended. Bit convoluted, maybe you'll find something more elegant than this. And maybe we should use `VALUE_NONE` in 4.0, to simplify all this.

agustingomes committed:
Apply workaround to maintain expected behavior

This test case addition simulates what happens when the
`--all-or-nothing` option is indicated, but no explicit
value is passed.

We'll set a default option, which is retrieved in case the option
has not been provided, allowing us to override the value in case the
option is different than the default value.

agustingomes commented:
> And maybe we should use VALUE_NONE in 4.0, to simplify all this.

That sounds like a good idea.

greg0ire commented:
You can use `vendor/bin/phpcbf` to fix the cs issues.

agustingomes commented:
@greg0ire how do you want to proceed regarding the PHPStan failures? THey seem to be originating from deprecations, and the BC check error I never seen it before.

greg0ire commented on lib/Doctrine/Migrations/Tools/Console/ConsoleInputMigratorConfigurationFactory.php:
```diff
@@ -21,12 +22,26 @@ public function getMigratorConfiguration(InputInterface $input): MigratorConfigu
+    {
+        try {
+            $allOrNothingOption = $input->getOption('all-or-nothing');
+        } catch (InvalidArgumentException $exception) {
```
For which cases is this necessary? I have an uneasy feeling about this… if it's not strictly related to the issue at hand, I'd remove it.

agustingomes replied:
There are some tests that fail if we don't catch it... I found it weird as well, but I would not know how to proceed further otherwise.

The tests that failed when not catching the exception:

- [testWriteSql](https://github.com/doctrine/migrations/blob/3.5.x/tests/Doctrine/Migrations/Tests/Tools/Console/Command/ExecuteCommandTest.php#L50)
- [testExecute](https://github.com/doctrine/migrations/blob/3.5.x/tests/Doctrine/Migrations/Tests/Tools/Console/Command/ExecuteCommandTest.php#L101)
- [testExecuteMultiple](https://github.com/doctrine/migrations/blob/3.5.x/tests/Doctrine/Migrations/Tests/Tools/Console/Command/ExecuteCommandTest.php#L123)
- [testExecuteCancel](https://github.com/doctrine/migrations/blob/3.5.x/tests/Doctrine/Migrations/Tests/Tools/Console/Command/ExecuteCommandTest.php#L155)

greg0ire replied:
I think this class you're changing is called for several commands. IMO you should remove this `try/catch`, and resort to `hasOption` here, to determine if the command that is called has that option (as in, the option is defined as opposed to its value provided). If it doesn't, then I guess we can omit calling `setAllOrNothing`, or call it with `false`, whatever is the most simple for you.

agustingomes replied:
The last push I did seems to be passing all tests. Using the `hasOption` did help.

greg0ire commented:
Nice commit messages. I've seen the test failing, so you can squash both commits now, so that all commits pass the build.

greg0ire commented:
> @greg0ire how do you want to proceed regarding the PHPStan failures? THey seem to be originating from deprecations, and the BC check error I never seen it before.

Don't worry about those, I'll take care of it.

Regarding the BC-check, it seems to have been broken for a while: https://github.com/Nyholm/roave-bc-check-docker/issues/32
Regarding the static analysis issues, I created https://github.com/doctrine/migrations/pull/1297

greg0ire commented on lib/Doctrine/Migrations/Tools/Console/ConsoleInputMigratorConfigurationFactory.php:
```diff
@@ -21,12 +21,26 @@ public function getMigratorConfiguration(InputInterface $input): MigratorConfigu
+            return null;
+        }
+
+        return (bool) ($allOrNothingOption ?? true);
```
If the command does not have the option, the value will default to `true`, but I guess it does not matter since `all-or-nothing` is not used when the option is not defined.

e-jevdokimovs replied:
This is a problem, the default value of `MigrationConfiguration` property `allOrNothing` is `false`, these changes always set the value to true if the option isn't passed at all AND the command doesn't have an `all-or-nothing` option, for example `ExecuteCommand`.

greg0ire replied:
You mean the option is used although it's not defined in the command? Where?

e-jevdokimovs replied:
The option itself isn't used, but [ConsoleInputMigratorConfigurationFactory->getMigratorConfiguration()](https://github.com/doctrine/migrations/blob/3.5.x/lib/Doctrine/Migrations/Tools/Console/ConsoleInputMigratorConfigurationFactory.php#L20) is being called in [`ExecuteCommand->execute()`](https://github.com/doctrine/migrations/blob/3.5.x/lib/Doctrine/Migrations/Tools/Console/Command/ExecuteCommand.php#L115) which in turn creates a `MigratorConfiguration` with `allOrNothing` set to `true` because `ExecuteCommand` doesn't have an `all-or-nothing` option which results in the value not defaulting to `notprovided`.

e-jevdokimovs replied:
Basically this method should look like this:
```
        if (!$input->hasOption('all-or-nothing')) {
            return null;
        }

        $allOrNothingOption = $input->getOption('all-or-nothing');

        if ($allOrNothingOption === 'notprovided') {
            return null;
        }

        return (bool) ($allOrNothingOption ?? true);
```

e-jevdokimovs replied:
Created a PR with the fix: https://github.com/doctrine/migrations/pull/1308

greg0ire commented:
Thanks @agustingomes !

greg0ire commented:
> > And maybe we should use VALUE_NONE in 4.0, to simplify all this.
> 
> That sounds like a good idea.

@agustingomes if you want to work on this, you can open a PR on 3.6.x to deprecate passing a value :)

Here is an example: https://github.com/doctrine/migrations/blob/01f89a1082b6971ba53ba76f95741fc8ad4a5fd2/lib/Doctrine/Migrations/Tools/TransactionHelper.php#L22-L32

agustingomes commented:
@greg0ire I'll look into it the next days 👍🏼
